fv3gfs-util/fv3gfs/util/partitioner.py

```diff
@@ -712,39 +809,136 @@ def tile_extent_from_rank_metadata(
+        that lets the interior tile sizes still be an integer. After that, it counts up from the starting edge size.
+        """
+        bottom = 1
+        top = int((dim_extent - subtile_count + 2) / 2) + 1
+        unsorted_valid_sizes = range(bottom, top)
+        valid_sizes = []
+
+        index = start
+        offset = 0
+        factor = -1
+
+        for i in range(len(unsorted_valid_sizes) + 1):
+            index = start + factor * offset
+            if index in unsorted_valid_sizes and index not in valid_sizes:
+                valid_sizes.append(index)
+            offset = offset + 1
+            if index == 1:
+                offset = 0
+                factor = 1
+        return valid_sizes
+
+    layout_factors = np.asarray(
         utils.list_by_dims(dims, layout, non_horizontal_value=1)
     )
-    return extent_from_metadata(dims, tile_extent, layout_factors)
+
+    return_extents = []
+    edge_extents = []
+    for dim, subtile_count, dim_extent in zip(dims, layout_factors, tile_extent):
+        dim_edge_interior_ratio = edge_interior_ratio
+        if dim in constants.INTERFACE_DIMS:
+            dim_extent = dim_extent - 1
+        if (not subtile_count % 2) and dim_extent % 2:
+            raise ValueError(
+                f"Cannot find valid decomposition for odd ({dim_extent}) gridpoints along an even count ({subtile_count}) of ranks."
+            )
+        if (
+            subtile_count >= 3 and dim in constants.HORIZONTAL_DIMS
+        ):  # only do shrinked edges in x,y and if there is interior
+            edge_subtile_size = round(
+                dim_extent / (2 + (subtile_count - 2) / dim_edge_interior_ratio)
+            )
+
+            # searching of a valid integer pair for edge and interior tile sizes that add up to the entire dimension extent.
+            found = False
+            for edge_size in _valid_edge_tile_sizes(
+                dim_extent, subtile_count, edge_subtile_size
+            ):
+                dim_edge_interior_ratio = edge_size / (
+                    (dim_extent - 2 * edge_size) / (subtile_count - 2)
+                )
+                if (
+                    edge_size * 2
+                    + (subtile_count - 2) * int(edge_size / dim_edge_interior_ratio)
+                    == dim_extent
+                ):
+                    found = True
+                    break
+            if not found:
+                raise ValueError(
+                    f"No valid subdomain assignment found for dimension {dim} with {dim_extent} gridpoints along {subtile_count} ranks."
+                )
+            return_extents.append(int(edge_size / dim_edge_interior_ratio))
+            edge_extents.append(int(edge_size))
+        else:
+            edge_size = int(dim_extent / subtile_count)
+            return_extents.append(edge_size)
+            edge_extents.append(edge_size)
+    return_extents.extend(edge_extents)
```

Comment:
This is pretty long and complex. In some cases I'd ask the code author to refactor this into smaller bits to be clearer. Instead can you add a description somewhere, either as a multi-line comment or as comments throughout the code, of the general algorithm you've implemented here for solving the problem?

Reply:
I've added some in-line comments to this block of code.